swarm/storage/localstore/db.go

```diff
@@ -0,0 +1,230 @@
+package localstore
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/swarm/shed"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+/*
+   types of access:
+   - just get the data
+   - increment access index
+
+   - when uploaded or pull synced
+   - when delivered
+   - when push synced
+   - when accessed
+*/
+
+var (
+	errInvalidMode = errors.New("invalid mode")
+)
+
+// Modes of access/update
+const (
+	SYNCING rushed.Mode = iota
+	UPLOAD
+	REQUEST
+	SYNCED
+	ACCESS
+	REMOVAL
+)
+
+// DB is a local chunkstore using disk storage
+type DB struct {
+	*rushed.DB
+	// fields and indexes
+	schemaName shed.StringField
+	size       shed.Uint64Field
+	retrieval  shed.Index
+	push       shed.Index
+	pull       shed.Index
+	gc         shed.Index
+}
+
+// NewDB constructs a local chunks db
+func NewDB(path string) (*DB, error) {
+	db := new(DB)
+	sdb := shed.NewDB(path)
+	db.DB = rushed.NewDB(sdb, db.update, db.access)
+	db.schemaName, err = idb.NewStringField("schema-name")
+	if err != nil {
+		return nil, err
+	}
+	db.size, err = idb.NewUint64Field("size")
+	if err != nil {
+		return nil, err
+	}
+	db.retrieval, err = idb.NewIndex("Hash->StoredTimestamp|AccessTimestamp|Data", shed.IndexFuncs{
+		EncodeKey: func(fields shed.IndexItem) (key []byte, err error) {
+			return fields.Hash, nil
+		},
+		DecodeKey: func(key []byte) (e shed.IndexItem, err error) {
+			e.Hash = key
+			return e, nil
+		},
+		EncodeValue: func(fields shed.IndexItem) (value []byte, err error) {
+			b := make([]byte, 16)
+			binary.BigEndian.PutUint64(b[:8], uint64(fields.StoreTimestamp))
```

Comment:
Any reason why we don't just implement the [`BinaryMarshaler`](https://golang.org/pkg/encoding/#BinaryMarshaler) interface?

Reply:
I think we don't use Go interfaces enough. One of the key features of Go are interfaces and they make the code really flexible.

That goes as far as I think EncodeKey and DecodeKey should be binary marshal and unmarshal on an IndexKey and same goes for IndexValue. The IndexIteam could be a composition of the key and value types.

I believe that could result in a more idiomatic Go code and could make the code look immediately familiar to any Gopher.

Reply:
This comment is pore related how `shed` is designed, and I would like to see your thoughts on https://github.com/ethersphere/go-ethereum/pull/993. There is a go example test where the usage is demonstrated.

The decision not to use interfaces for encode/decode functions is to avoid performance penalties when using them by introducing abstractions. Interfaces are a bit costly and allocate to heap. Encode/decode functions will be called all the time in swarm, as the storage is lower layer. This is the reasoning why IndexItem is implemented in most simple value semantics, instead interfaces. I tried to document this in comments in the PR https://github.com/ethersphere/go-ethereum/pull/993, and would like to hear other opinions. If needed, I can do another demonstration, maybe in more detail then I did last week where I talked about this decisions.

Reply:
@janos 
The code this complex I have to sit down and read, that's just how my brain works. So I'll go trough shed, too. Thanks for the demonstration offer though. :)

If you have some good articles in mind, our your own measurements about these performance benefits I would be happy to see/read. If not, I'll google. :)

Reply:
The most simple demonstration how interfaces are complied, escaped or inlined is here https://github.com/golang/go/issues/20116#issuecomment-297140689. But it just boils down to how interfaces in implemented in go and how compiler decides what to leave to runtime. 

If you are interested in implementing encode/decode functions with interfaces, I would be glad to see benchmarks and how the example will look like with them as we would need additional types in example/usage instead anonymous functions. Performance was the main question after the first demonstration and for the right reason, as this part of the code is the hot path.

Reply:
Or maybe we can just leave the performance considerations aside as they may be neglectable compared to database operations. :)

Reply:
I guess this will be a good start for me: https://github.com/golang/go/issues/20116 :)

Reply:
Also, I guess as to compiler improves some of these penalties will disappear: 
https://github.com/golang/go/issues/19361

Reply:
> Or maybe we can just leave the performance considerations aside as they may be neglectable compared to database operations. :)

I guess that's also possible. Perf: never guess. :)

What I was thinking about: Are we optimizing prematurely?

Would it be better 1st write the code with interfaces then optimize?
Considerations:
- Maybe the code ends up more ideal in look and we have to tune only a few places.
- What about optimizing for change? ([XP](https://en.wikipedia.org/wiki/Extreme_programming) core value) Change is also a cost and imminent. 
- If we first go with interfaces, then what would be the blast radius (cost) of replacing them? Given that we try to expose/export only as much as required.

I would vote for interfaces, but I have limited experience here. So I have to dig deeper.

Reply:
Great, this is the discussion that I was looking for. I would be glad to hear other opinions too. If we are up to use interfaces, then struct fields also can help in simplifying usage by applying declarative approach instead of the current functional one.